src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -738,17 +738,31 @@ RunLocalExecution(CitusScanState *scanState, DistributedExecution *execution)
 static void
 AdjustDistributedExecutionAfterLocalExecution(DistributedExecution *execution)
 {
-	/*
-	 * Local execution already stores the  tuples for returning, so we should not
-	 * store again.
-	 */
-	execution->hasReturning = false;
-
 	/* we only need to execute the remote tasks */
 	execution->tasksToExecute = execution->remoteTaskList;
 
 	execution->totalTaskCount = list_length(execution->remoteTaskList);
 	execution->unfinishedTaskCount = list_length(execution->remoteTaskList);
+
+	/*
+	 * For reference tables, local execution already stores the  tuples for
+	 * returning, so we should not store again.
+	 *
+	 * We typically expect to have modifications to reference tables when the
+	 * unfinishedTaskCount is one. However, we still explicitly check for
+	 * reference tables because there are some edge cases for distributed
+	 * tables with 2 shards.
```

Comment:
What are the edge cases?

Reply:
```SQL


SELECT start_metadata_sync_to_node(nodename, nodeport) FROM pg_dist_node;

SET citus.shard_replication_model TO 'streaming';

SET citus.shard_count TO 2;

CREATE TABLE test_table (a INT);
SELECT create_distributed_table('test_table', 'a');


-- connect to a worker
psql -p 9700

-- do a remote insert
INSERT INTO test_table VALUES (2);

BEGIN;
	-- do a local insert
	INSERT INTO test_table VALUES (1);

    -- if we don't check for reference tables, this would skip returning 2
	DELETE FROM test_table RETURNING *;
```